yarn-project/prover-client/src/job_pool/broker.ts

```diff
@@ -0,0 +1,219 @@
+  lastUpdatedAt: number;
+};
+
+type ProofRequestBrokerConfig = {
+  timeoutIntervalMs?: number;
+  proofRequestTimeoutMs?: number;
+  maxRetries?: number;
+};
+
+/**
+ * A broker that manages proof requests and distributes them to workers based on their priority.
+ * It takes a backend that is responsible for storing and retrieving proof requests and results.
+ */
+export class ProofRequestBroker implements ProofRequestProducer, ProofRequestConsumer {
+  private queues: QueueByProofType = Object.values(ProofType).reduce((acc, proofType) => {
+    acc[proofType] = new PriorityMemoryQueue<ProofRequest<typeof proofType>>(proofPriorityComparator) as any;
+    return acc;
+  }, {} as QueueByProofType);
+
+  private inProgress = new Map<ProofRequestId, InProgressMetadata>();
+  private retries = new Map<ProofRequestId, number>();
+
+  private timeoutPromise: RunningPromise;
+  private timeSource = Date.now;
+  private proofRequestTimeoutMs: number;
+  private maxRetries: number;
+
+  public constructor(
+    private backend: BrokerBackend,
+    { proofRequestTimeoutMs = 30_000, timeoutIntervalMs = 10_000, maxRetries = 3 }: ProofRequestBrokerConfig = {},
+    private logger = createDebugLogger('aztec:prover-client:proof-request-broker'),
+  ) {
+    this.timeoutPromise = new RunningPromise(this.timeoutCheck, timeoutIntervalMs);
+    this.proofRequestTimeoutMs = proofRequestTimeoutMs;
+    this.maxRetries = maxRetries;
+  }
+
+  public async start(): Promise<void> {
+    for (const item of this.backend.allProofRequests()) {
+      this.logger.info(`Re-enqueueing proof request id=${item.id} type=${item.proofType}`);
+      this.enqueueProofInternal(item);
+    }
+
+    this.timeoutPromise.start();
+  }
+
+  public stop(): Promise<void> {
+    return this.timeoutPromise.stop();
+  }
+
+  public async enqueueProof<T extends ProofType>(request: ProofRequest<T>): Promise<void> {
+    await this.backend.saveProofRequest(request);
+    return this.enqueueProofInternal(request);
+  }
+
+  public cancelProof(id: ProofRequestId): Promise<void> {
+    this.logger.info(`Cancelling proof request id=${id}`);
+    this.retries.delete(id);
+    this.inProgress.delete(id);
+    return this.backend.removeProofRequest(id);
+  }
+
+  public async getProofStatus<T extends ProofType>(id: ProofRequestId, proofType: T): Promise<ProofRequestStatus<T>> {
+    const item = this.backend.getProofRequest(id, proofType);
+    if (!item) {
+      this.logger.warn(`Proof request id=${id} type=${proofType} not found`);
+      return Promise.resolve({ status: 'not-found' });
+    }
+
+    const result = this.backend.getProofResult(id, proofType);
+    if (!result) {
+      return Promise.resolve({ status: this.inProgress.has(id) ? 'in-progress' : 'in-queue' });
+    } else if ('value' in result) {
+      return Promise.resolve({ status: 'resolved', value: result.value });
+    } else {
+      return Promise.resolve({ status: 'rejected', error: result.error });
+    }
+  }
```

Comment:
Eventually, do you think we could respond to this request just by hitting in-memory structures? I'm assuming we keep the backend just for recovering from crashes, or for sharing payloads, as mentioned in the other above.

Reply:
It's now only hitting the in memory cache. The function itself still has to be async in order for the JSON-RPC proxy to work properly with it.